cbcdb/main.py

```diff
@@ -0,0 +1,375 @@
+
+    def _database_connection_sub_method(self, host, port, method_instance, sql, params):
+        with connect(dbname=self._db_name,
+                     host=host,
+                     port=port,
+                     user=self._db_user,
+                     password=self._db_password) as conn:
+            with conn.cursor() as curs:
+                return method_instance(sql, params, curs, conn)
+
+    def get_sql_dataframe(self, sql: str, params: list = None, curs=False, conn=False) -> pd.DataFrame:
+        """
+         Returns a DataFrame for a given SQL query
+
+        Args:
+            sql: SQL string
+            params: List of parameters
+            execute_flag: If set to false, the sql query will call the connection method to open tunnel and make the
+                          database connection. The connection method will recursively call this method with the
+                          execute_flag set to true.
+
+        Returns: A Pandas Dataframe.
+
+        """
+        self._print_debug_output(f"Getting query:\n {sql}")
+
+        if curs:
+            if params:
+                df = pd.read_sql_query(sql, params=params, con=conn)
+            else:
+                df = pd.read_sql_query(sql, con=conn)
+        else:
+            df = self._get_connection(sql, params, self.get_sql_dataframe)
+        return df
+
+    def get_sql_list_dicts(self, sql: str, params: list = None, curs=False, conn=False) -> List[Dict[str, Any]]:
+        """
+        Returns a list of dicts for a given SQL Query
+
+        Args:
+            sql: SQL string
+            params: Parameters for SQL call
+            curs: A cursor instance. If present the method will execute the SQL. If set to False, the method will
+            call the connection method. The connection method will recursively call this method with the cursor.
+
+
+        Returns:
+            Example Output:
+            [{'some': 'data'},
+             {'more': 'otherdata'}]
+
+        """
+        self._print_debug_output(f"Getting query:\n {sql}")
+        if curs:
+            if params:
+                curs.execute(sql, params)
+            else:
+                curs.execute(sql)
+            output = []
+            columns = [column[0] for column in curs.description]
+            for row in curs.fetchall():
+                output.append(dict(zip(columns, row)))
+        else:
+            output = self._get_connection(sql, params, self.get_sql_list_dicts)
+        return output
+
+    def get_sql_single_item_list(self, sql: str, params: list = None, curs=False, conn=False) -> list:
+        """
+        Returns a single column list for a given SQL Query
+
+        Args:
+            sql: SQL string
+            params: List of parameters
+
+        Returns: A list containing the results of the query
+        """
+        self._print_debug_output(f"Getting query:\n {sql}")
+
+        if curs:
```

Comment:
The more I look at this, I would try to extract all the checks for `curs` and the conditions around that into a method that executes the sql (and obtains a cursor if one doesn't already exist).  Then pretty much all of these methods only need to consist of a call to that method, and then the necessary processing of the result to put it in the correct data structure.

Reply:
To be more specific, I think `_database_connection_sub_method` should be modified to simply take in the sql statement, execute it, and return the result. Then let the calling method handle the processing into the proper data structure. I don't think the method needs to be passed down to `_database_connection_sub_method`

Reply:
Let me know if that's clear or needs more explanation, not sure I'm stating that well enough